open_spiel/games/gin_rummy.cc

```diff
@@ -0,0 +1,699 @@
+    legal_actions.push_back(kKnockAction);
+  }
+  std::sort(legal_actions.begin(), legal_actions.end());
+  return legal_actions;
+}
+
+std::vector<Action> GinRummyState::KnockLegalActions() const {
+  std::vector<Action> legal_actions;
+  // After knocking, the player discards. This discard must not prevent
+  // the player from arranging the hand in such a way that the deadwood
+  // total is less than the knock card.
+  if (hands_[cur_player_].size() == 11) {
+    std::vector<int> legal_discards =
+        LegalDiscards(hands_[cur_player_], knock_card_);
+    for (int card : legal_discards) {
+      legal_actions.push_back(card);
+    }
+  } else {
+    std::vector<int> meld_ids = LegalMelds(hands_[cur_player_], knock_card_);
+    for (int meld_id : meld_ids) {
+      int action_id = meld_id + kMeldActionBase;
+      legal_actions.push_back(action_id);
+    }
+    // Must keep laying melds until remaining deadwood is less than knock card.
+    // Once that has been accomplished, the knocking player can pass.
+    if (TotalCardValue(hands_[cur_player_]) <= knock_card_) {
+      legal_actions.push_back(kPassAction);
+    }
+  }
+  std::sort(legal_actions.begin(), legal_actions.end());
+  return legal_actions;
+}
+
+std::vector<Action> GinRummyState::LayoffLegalActions() const {
+  std::vector<Action> legal_actions;
+  if (!finished_layoffs_) {
+    // Always have the option not to lay off any cards.
+    legal_actions.push_back(kPassAction);
+    std::vector<int> all_possible_layoffs =
+        AllLayoffs(layed_melds_[prev_player_], layoffs_);
+    for (int card : all_possible_layoffs) {
+      if (std::find(hands_[cur_player_].begin(),
+                    hands_[cur_player_].end(), card) !=
+                        hands_[cur_player_].end()) {
+        legal_actions.push_back(card);
+      }
+    }
+  } else {
+    // After laying off individual cards, now the player lays melds.
+    // Always have the option not to declare any melds.
+    legal_actions.push_back(kPassAction);
+    // The non-knocking player does not have to arrange melds in a particular
+    // way to get under the knock card. Therefore we use kMaxPossibleDeadwood
+    // to ensure that all melds are legal.
+    std::vector<int> meld_ids = LegalMelds(hands_[cur_player_],
+                                           kMaxPossibleDeadwood);
+    for (int meld_id : meld_ids) {
+      int action_id = meld_id + kMeldActionBase;
+      legal_actions.push_back(action_id);
+    }
+  }
+  std::sort(legal_actions.begin(), legal_actions.end());
+  return legal_actions;
+}
+
+std::vector<Action> GinRummyState::WallLegalActions() const {
+  std::vector<Action> legal_actions;
+  legal_actions.push_back(kPassAction);
+  // Player can either pass or knock (if legal).
+  int deadwood = MinDeadwood(hands_[cur_player_], upcard_);
+  if (deadwood <= knock_card_) {
+    legal_actions.push_back(kKnockAction);
+  }
+  std::sort(legal_actions.begin(), legal_actions.end());
```

Comment:
given you are only adding two actions and they're constants, can you add them in sorted order so you can either remove the sort or at least make it a no-op? If that's already true, might be worth a comment.

Reply:
Removed sort in new commit.